[BUGFIX] - Merge query params from previous transition into active transition

raytiley opened the pull request:
This merges in queryParams from an active transition in such a way that `prepareQueryParams` won't duplicate them.

It also won't merge in the qp if it is already int he qp changed list, for example if it's already been set during route setup by a call to `controller.set('qpProp', 'someValue')`.

Fixing these things with my current test jsbin also recreated another issue that has been biting people and that is that a call to `transitionTo` creates an aborted transition that calls `didTransition` with an empty set of handlerInfos. This would cause the `updatePaths` method to fail since it couldn't calculate the paths without the handler infos. We now bail early from `updatePaths` it is called with an empty array of handlerInfos.

I think this is fine because the aborted transition is followed by a successful transition that will update the paths appropriatly.

Still todo - TESTS

raytiley committed:
Normalize query params when merging in from active transition

This merges in queryParams from an active transition in such a way that
`prepareQueryParams` won't duplicate them.  It also won't merge in the
qp if it is already int he qp changed list, for example if it's already
been set during route setup by a call to `controller.set('qpProp',
'someValue')`.  Fixing these things with my current test jsbin also
recreated another issue that has been biting people and that is that a
call to `transitionTo` creates an aborted transition that calls
`didTransition` with an empty set of handlerInfos. This would cause the
`updatePaths` method to fail since it couldn't calculate the paths
without the handler infos. We now bail early from `updatePaths` it is
called with an empty array of handlerInfos.  I think this is fine
because the aborted transition is followed by a successful transition
that will update the paths appropriatly.

This gets the active queryParams and the provided using the same key
`controller:prop` so that the calls to `assign` override as expected.
This fixes issues where the QP doesn't update properly on refresh
because the merged value conflicts with the provided value.

rwjblue commented on packages/ember-routing/lib/system/router.js:
```diff
@@ -636,7 +636,17 @@ var EmberRouter = EmberObject.extend(Evented, {
     // queryparams from the url on initial load.
     if (this.router.activeTransition) {
-      assign(queryParams, this.router.activeTransition.queryParams);
+      var unchangedQPs = {};
```
Can this bit of code live in a separate method (`_processActiveTransitionQueryParams` or something)?

rwjblue commented:
Thanks for working on this @raytiley, let me know if I can help...

raytiley commented:
Come to Maine and babysit for a few hours?  I should be able to wrap this
up tonight.

On Sunday, April 10, 2016, Robert Jackson <notifications@github.com> wrote:

> Thanks for working on this @raytiley <https://github.com/raytiley>, let
> me know if I can help...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/emberjs/ember.js/pull/13273#issuecomment-208006582>
>

rwjblue commented:
> Come to Maine and babysit for a few hours? I should be able to wrap this up tonight.

Sounds fun! Might have to make a pit stop at the Bissell Brothers though....

knownasilya commented:
@raytiley I'm up to help any way I can, although visiting to babysit might be out of the question.

igorT commented on packages/ember-routing/lib/system/router.js:
```diff
@@ -694,6 +704,33 @@ var EmberRouter = EmberObject.extend(Evented, {
+      for (var j = 0, qpLen = qpMeta.qps.length; j < qpLen; ++j) {
+        var qp = qpMeta.qps[j];
+
+        var presentProp = qp.prop in queryParams  && qp.prop ||
```
Can you please break this down and add more local vars to make this more readable

raytiley commented:
Hey Everyone... sorry. I've been really busy at work. I promise I will clean this up / get tests this week.

igorT commented:
I've been testing this PR in our app and it seems to fix certain query param issues that broke in 2.4.4 but introduces another bug where initial load of a page with a query 404s. Investigating

igorT commented on packages/ember-routing/lib/system/router.js:
```diff
@@ -694,6 +704,33 @@ var EmberRouter = EmberObject.extend(Evented, {
+  _normalizeQueryParams(leafRouteName, contexts, queryParams) {
+    var state = calculatePostTransitionState(this, leafRouteName, contexts);
+    var handlerInfos = state.handlerInfos;
+    var appCache = this._bucketCache;
```
This is unused

raytiley commented:
First steps... I rebased the code and fixed the jshint error. Tomorrow I'll write some tests based on some jsbin cases I used to write this code. Sorry for the delay.

Commit:
Add renetrant test for query params changes

igorT commented:
@raytiley I added some tests here for things that worked in 2.4.3 and broke in 2.4.4 I think your branch fixed them as of yesterday. 
https://github.com/emberjs/ember.js/commit/53fcdbc6376981d8a219cde00d04bd18d59cac9d

pixelhandler commented on packages/ember/tests/routing/query_params_test.js:
```diff
@@ -2490,6 +2490,43 @@ if (isEnabled('ember-routing-route-configured-query-params')) {
     bootApplication();
   });
 
+  QUnit.test('queryParams are updated when a controller property is set and the route is refreshed. Issue #13263  ', function() {
```
This test looks good!

Seems to me that this test will fail with "Uncaught Error, Assertion Failed: You're not allowed to have more than one controller property map to the same query param key…"  without the code to fix.

lolmaus commented:
This issue needs to be properly tagged. At least with `Query Params`.

lolmaus commented:
@Sinled has discovered a regression with an array query param.

Previously, you could have an array query param like this:

```js
export default Ember.Controller.extend({
  queryParams: ['fields',],
  fields: ['field1', 'field2']
})
```

Ember would properly (and transparently!) serialize them into strings and then deserialze back.

After the regression, array query params are not deserialized and the Controller's array property gets rewritten with a string.

Me and @Sinled managed to narrow down the regression to this commit by @raytiley: `master` https://github.com/emberjs/ember.js/commit/2b42d5659498df2bf725f09e65632bfe5c3161c0, `lts-2-4` https://github.com/emberjs/ember.js/commit/7460b686d67eb349068482ddad131cd3c76c3dc1.

Please tell us whether:

* this is a bug or works as designed;
* this is a known or new issue;
* it should be reported as a new GitHub issue ticket.

rwjblue commented:
> This issue needs to be properly tagged. 

This comment is not helpful, please keep things constructive.

> Please tell us whether:
> 
> * this is a bug or works as designed;

Bug.

> * this is a known or new issue;

If you didn't find it via a GitHub search, it is a new issue.  I personally do not recall an issue opened about 2.4.4 regarding array query params.

> * it should be reported as a new GitHub issue ticket.

If you didn't find it via a GitHub search, it is a new issue.  Please test this PR to see if it fixes.

mattmarcum commented on packages/ember-routing/lib/system/router.js:
```diff
@@ -987,6 +1032,8 @@ function calculatePostTransitionState(emberRouter, leafRouteName, contexts) {
 
 function updatePaths(router) {
   let infos = router.router.currentHandlerInfos;
+  if (infos.length === 0) { return; }
```
I'm running into a similar problem but in my case `infos` isn't even an array.  I did something like:

```
import { isArray } from 'ember-runtime/utils';
...
if (!isArray(infos) || infos.length === 0) { return; }
```

homu commented:
:umbrella: The latest upstream changes (presumably b125ff5) made this pull request unmergeable. Please resolve the merge conflicts.

knownasilya commented:
I'm totally getting bit by the QPs not being deserialized properly (arrays) in the `queryParamsDidChange` hook. How's this PR going, and would it solve that?

pixelhandler commented:
@stefanpenner @chancancode question for you, do we need a "regression" label on this?

pixelhandler commented:
@raytiley thanks for putting this PR into play. This will help people that would like to adopt the LTS release. I spoke to a few people at our meetup Wed. And noticed those interested in LTS are running 2.4.3 instead of 2.4.5 due to #13258

lolmaus commented:
Related: https://github.com/emberjs/ember.js/issues/13591

Sinled commented:
2.4.6 fixed my [problem](https://github.com/emberjs/ember.js/pull/13273#issuecomment-214739521l) with query params serialisation, but it still present in 2.6.0

rwjblue commented:
Confirm. This PR is still needed. 

@igorT - Were you going to pick this up?

homu commented:
:umbrella: The latest upstream changes (presumably #13658) made this pull request unmergeable. Please resolve the merge conflicts.

rwjblue commented:
🎉

nathanhammond commented:
Hey everybody who paid attention to this thread! We've a few open bugs likely triggered by this change. Anybody have time to dive in?
